### PR TITLE
DEV: Plugin API to add directory columns (take 2)

### PR DESCRIPTION
**Pull request by markvanlan**
@eviltrout The new parts here compared to the version of this that was merged last week is:

* Directory columns loaded via ajax rather than preload store
* Changed plugin API so that the `DirectoryColumn` record is added during `DirectoryItem#refresh` rather than inside the plugin api

### CODE REVIEW

File: app/assets/javascripts/discourse/app/routes/users.js [hunk 1]
```diff
@@ -38,9 +39,12 @@ export default DiscourseRoute.extend({
   },
 
   model(params) {
-    const columns = PreloadStore.get("directoryColumns");
-    params.order = params.order || columns[0].name;
-    return { params, columns };
+    return ajax("directory-columns.json")
```
Comment (markvanlan) [inline app/assets/javascripts/discourse/app/routes/users.js:42]: The previous PR I did for this, used the preload store. This is now an ajax request

File: app/assets/javascripts/discourse/tests/helpers/create-pretender.js [hunk 2]
```diff
@@ -1035,4 +1035,75 @@ export function applyDefaultHandlers(pretender) {
       ],
     });
   });
+
+  pretender.get("directory-columns.json", () => {
```
Comment (markvanlan) [inline app/assets/javascripts/discourse/tests/helpers/create-pretender.js:1039]: No preload store anymore. Now we have this like I mentioned above

File: app/controllers/directory_columns_controller.rb [hunk 3]
```diff
@@ -1,15 +1,20 @@
...
 class DirectoryColumnsController < ApplicationController
-  requires_login
+  requires_login except: [:index]
 
   def index
+    directory_columns = DirectoryColumn.includes(:user_field).where(enabled: true).order(:position)
+    render_json_dump(directory_columns: serialize_data(directory_columns, DirectoryColumnSerializer))
+  end
+
+  def edit_index
```
Comment (markvanlan) [inline app/controllers/directory_columns_controller.rb:11]: Kinda weird method name here... but we have 2 indexes. One is public index and one is edit index so I went with it.

Comment (eviltrout) [reply app/controllers/directory_columns_controller.rb:11]: This is a sign that you should create a new controller. It maps to your URLs below too. You could have `DirectoryColumnsController` and `EditDirectoryColumnsController`.

File: spec/components/plugin/instance_spec.rb [hunk 4]
```diff
@@ -600,4 +600,55 @@ def register_locale(locale, opts)
...
+        plugin.add_directory_column(column_name, query: "SELECT COUNT(*) FROM users", icon: 'recycle')
+        expect(DirectoryColumn.find_by(name: column_name, icon: 'recycle', enabled: false).present?).to be(false)
+
+        DirectoryItem.refresh!
+        expect(DirectoryColumn.find_by(name: column_name, icon: 'recycle', enabled: false).present?).to be(true)
+
+        # add_directory_column runs `DiscourseEvent.on("before_directory_refresh")`. The rails spec helper
+        # ensures that no test will run and change the total number of registrations. The easy solution is to
+        # manually clean up the registration:
+        DiscourseEvent.events.delete("before_directory_refresh")
```
Comment (eviltrout) [inline spec/components/plugin/instance_spec.rb:636]: This feels a bit messy to me. Can you explain more about what you mean by "ensures that no test will run and change the total number of registrations". I would really prefer if we didn't have to delete stuff for tests to pass.

Comment (markvanlan) [reply spec/components/plugin/instance_spec.rb:636]: This is the error with the line removed
```
1) Plugin::Instance#add_directory_column with valid column name creates a directory column record when directory items are refreshed
     Failure/Error: expect(before_event_count).to eq(after_event_count), "DiscourseEvent registrations were not cleaned up"
       DiscourseEvent registrations were not cleaned up
     # ./spec/rails_helper.rb:281:in `block (2 levels) in <top (required)>'
  
 ```
 
 Caused by https://github.com/discourse/discourse/blob/83a6ad32ffe75ae222028feddeca169fc5be54ac/spec/rails_helper.rb#L277-L282

Comment (eviltrout) [reply spec/components/plugin/instance_spec.rb:636]: I see, thanks. I would prefer if you weren't reaching into `events` to delete manually and we had a method for this.

Normally a test calls `.off` to remove an event it wires up. Perhaps you could add `.all_off` and call that?

Comment (markvanlan) [reply spec/components/plugin/instance_spec.rb]: Yeah, that's a really good solution. Pushed a commit that adds `DiscourseEvent#all_off`

**Review by eviltrout — APPROVED** [review]
thanks, looks good now.

**Pull request merged**